Firestore/core/src/firebase/firestore/model/blob.h

```diff
@@ -0,0 +1,34 @@
+/*
+ * Copyright 2018 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_MODEL_BLOB_H_
+#define FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_MODEL_BLOB_H_
+
+#include <cstdint>
+#include <vector>
+
+namespace firebase {
+namespace firestore {
+namespace model {
```

Comment:
Is it obvious that f:f::model::* is immutable? If not, consider adding a brief docstring to Blob stating that.

Reply:
I don't think I can quite make this work:
```c++
using Blob = const std::vector<uint8_t>;
```
But I can do
```c++
using Blob = std::shared_ptr<const std::vector<uint8_t>>;
```
WDYT?

Unfortunately this leaks out because now constructing the blob from raw bytes becomes:
```diff
-  Blob copy(source, source + size);
+  Blob copy = std::make_shared<Blob::element_type>(source, source + size);
```
Aside from this, it otherwise works.

Another possibility would just be to make a concrete Blob type that wraps this all up but is immutable.

Reply:
> I don't think I can quite make this work:
using Blob = const std::vector<uint8_t>;

A const reference might work, but then you still need to refer to... something/somewhere... which is what you've already come up with here:

> But I can do
using Blob = std::shared_ptr<const std::vector<uint8_t>>;

Yeah, I otherwise like this.

> Unfortunately this leaks out because now constructing the blob from raw bytes becomes:

Ah. Yeah, that's unfortunate. (Although, some other types use static factory methods to hide construction details. You *could* do that here too, though it obviously wouldn't prevent this usage.)

Another slight drawback is the 'Blob' (no star) is now used like a pointer.

> Another possibility would just be to make a concrete Blob type that wraps this all up but is immutable.

This is (I think?) where we'd want to eventually go as presumably, we're going to want to add some methods to the blob type (rather than just relying on whatever vector provides). eg. Blob.java provides things like toByteString(), toBytes(), toString(). It also allows us to not expose all vector methods.

So for this PR, I'm ok with just `using Blob = vector<uint8>` (optionally with a doc string stating it's "immutable" even though there's nothing enforcing that, and probably with a TODO with the following ->) but eventually, I think having a concrete type is where I'd naively expect this to end up. Going straight for the concrete type in this PR also seems perfectly reasonable and not that much more work.

Reply:
I started down the path of making this a concrete type but I think in the end there's just not enough value in it. Internally we're just going to wrap this up in a FieldValue already anyway so this just doubles the boilerplate required to make this a comparable, assignable thing.